src/nodes.coffee

```diff
@@ -1296,6 +1314,115 @@ exports.LineComment = class LineComment extends Base
 
 #### Call
 
+exports.CSXElement = class CSXElement extends Base
+  constructor: ({@tagName, @attributes, @content}) ->
+    super()
+
+  children: ['tagName', 'attributes', 'content']
+
+  compileNode: (o) ->
+    @attributes.base.csx = yes
+    @content?.base.csx = yes
+    fragments = [@makeCode('<')]
+    fragments.push (tag = @tagName.compileToFragments(o, LEVEL_ACCESS))...
+    if @attributes.base instanceof Arr
+      for obj in @attributes.base.objects
+        attr = obj.base
+        attrProps = attr?.properties or []
+        # Catch invalid CSX attributes: <div {a:"b", props} {props} "value" />
+        if not (attr instanceof Obj or attr instanceof IdentifierLiteral) or (attr instanceof Obj and not attr.generated and (attrProps.length > 1 or not (attrProps[0] instanceof Splat)))
+          obj.error """
+            Unexpected token. Allowed CSX attributes are: id="val", src={source}, {props...} or attribute.
+          """
+        obj.base.csx = yes if obj.base instanceof Obj
+        fragments.push @makeCode ' '
+        fragments.push obj.compileToFragments(o, LEVEL_PAREN)...
+    if @content
+      fragments.push @makeCode('>')
+      fragments.push @content.compileNode(o, LEVEL_LIST)...
+      fragments.push [@makeCode('</'), tag..., @makeCode('>')]...
+    else
+      fragments.push @makeCode(' />')
+    fragments
+
+  ast: ->
```

Comment:
While this is still "hairier" than most ASTs, I'd say this is a big improvement: we get to use the standard AST methods for constructing the outer `JSXElement` AST (getting rid of `Call::CSXToAst()`/`Call::CSXElementToAst()`)

Reply:
All the `compileNode` stuff that shows up as new code above is really only “new” in that it moved, right?

Reply:
Yes, exactly, it was `Call::compileCSX()`